File: 1475.final-prices-with-a-special-discount-in-a-shop.java
import java.util.Stack;

/*
 * @lc app=leetcode id=1475 lang=java
 *
 * [1475] Final Prices With a Special Discount in a Shop
 */

// @lc code=start
class Solution {
    public int[] finalPrices(int[] prices) {
        
        for(int i=0;i<prices.length-1;i++){
             for(int j=i+1;j<prices.length;j++){
            if(prices[j]<=prices[i]){
                prices[i]-=prices[j];
                break;
            }
        }
        }
        
        return prices;
    }
}
// @lc code=end

File: 1475.final-prices-with-a-special-discount-in-a-shop.py
#
# @lc app=leetcode id=1475 lang=python3
#
# [1475] Final Prices With a Special Discount in a Shop
#

# @lc code=start
from typing import List



class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        prices.reverse()
        stack = []
        result = []
        for price in prices:
            if not stack:
                result.append(0)
            elif stack[-1] < price:
                result.append(stack[-1])
            elif stack[-1] > price:
                while stack and stack[-1] > price:
                    stack.pop()
                if not stack:
                    result.append(0)
                else:
                    result.append(stack[-1])
            stack.append(price)
        result.reverse()
        print(result)
        return [(pri - res) for pri, res in zip(prices, result)]
            
        
# @lc code=end

